### PR TITLE
Add the x86 evaluators for rdbar/wrtbar for field watch

### PR DESCRIPTION
**Pull request by cathyzhyi**
Add evaluator supports for reporting field access.

issue: #2668

Signed-off-by: Yi Zhang <yizhang@ca.ibm.com>

### CODE REVIEW

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -116,6 +117,55 @@ static uint32_t logBase2(uintptrj_t n)
...
 
+static TR_OpaqueMethodBlock *
+getMethodFromBCInfo(
+      TR_ByteCodeInfo &bcInfo,
+      TR::Compilation *comp)
+   {
+   TR_OpaqueMethodBlock *method = NULL;
+
+   if (bcInfo.getCallerIndex() >= 0)
+      method = (TR_OpaqueMethodBlock *)(comp->getInlinedCallSite(bcInfo.getCallerIndex())._vmMethodInfo);
```
Comment (fjeremic) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: Curious why doesn't `getInlinedCallSite` just return `getCurrentMethod()` if index is -1?

Comment (cathyzhyi) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: the `_inlinedCallSites` is an array of TR_InlinedCallSiteInfo which encapsulate info about a callsite including `TR_InlinedCallSite` and bytecode info and some other queries that's only meaningful for an actual call node. Creating a dummy callsite for the top most method seem to be convenient for this particular code but would probably confuse a lot of other code expecting an actual callsite.

**Comment by cathyzhyi** [comment]
This PR  is dependent on https://github.com/eclipse/omr/pull/3027

**Comment by AdamBrousseau** [comment]
Jenkins copyright check

**Comment by AdamBrousseau** [comment]
Jenkins line endings check

**Comment by cathyzhyi** [comment]
@0dvictor could you help to review? Thanks!

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp [hunk 2]
```diff
@@ -12611,3 +12559,516 @@ J9::X86::TreeEvaluator::andORStringEvaluator(TR::Node *node, TR::CodeGenerator *
...
+      cg->recursivelyDecReferenceCount(sideEffectNode);
+   }
+
+TR::Register *J9::X86::TreeEvaluator::irdbarEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR::TreeEvaluator::rdbarSideEffectsEvaluator(node, cg);
+   if (!TR::Compiler->om.shouldGenerateReadBarriersForFieldLoads())
+      return TR::TreeEvaluator::iloadEvaluator(node, cg);
+
+   //sequence for pauseless GC
```
Comment (0dvictor) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: The official term is Concurrent Scavenge.

Comment (0dvictor) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: The official term for pause-less GC is Concurrent Scavenge.

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp [hunk 3]
```diff
@@ -12611,3 +12559,516 @@ J9::X86::TreeEvaluator::andORStringEvaluator(TR::Node *node, TR::CodeGenerator *
...
+   return TR::TreeEvaluator::dloadEvaluator(node, cg);
+   }
+
+TR::Register *J9::X86::TreeEvaluator::ardbarEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR::TreeEvaluator::rdbarSideEffectsEvaluator(node, cg);
+   if (!TR::Compiler->om.shouldGenerateReadBarriersForFieldLoads())
+      return TR::TreeEvaluator::aloadEvaluator(node, cg);
+
+   //sequence for pauseless GC
```
Comment (0dvictor) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: The official term for pause-less GC is Concurrent Scavenge.

**Comment by 0dvictor** [comment]
The changes looks good, just a comment update is needed as the official term for pause-less GC is Concurrent Scavenge.

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp (hunk 2 above)
Comment (cathyzhyi) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: changed the comment to the official term. Thanks!

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp (hunk 3 above)
Comment (cathyzhyi) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: changed the comment to the official term. Thanks!

**Comment by cathyzhyi** [comment]
waiting for jenkins builds pick up the depending omr change https://github.com/eclipse/omr/pull/3008

**Comment by cathyzhyi** [comment]
@fjeremic no, it's one-way dependency. current one depends on the merged one https://github.com/eclipse/omr/pull/3008 which added the `TR_EnableFieldWatch` option.

**Comment by fjeremic** [comment]
Jenkins test sanity all JDK8

**Comment by cathyzhyi** [comment]
@andrewcraik failures are JCL_Test_SE80_0/1 tests on all platforms. I think they are unrelated to this PR.

**Comment by fjeremic** [comment]
The failures are due to a cfg issue according to Peter:
https://openj9.slack.com/archives/CDS7QE9HB/p1541085437012100

I'll wait until the fix has propagated then kick off a new build to make sure.

**Comment by andrewcraik** [comment]
@fjeremic has the fix propagated now?

**Comment by fjeremic** [comment]
> @fjeremic has the fix propagated now?

Yes it has.

Jenkins test sanity all JDK8

**Comment by andrewcraik** [comment]
Mac failure is infrastructure related. The change should not impact OSX specifically so should be good to go.

**Pull request merged**